lib/src/Base/Optim/Bonmin.cxx

```diff
@@ -0,0 +1,312 @@
+ *
+ */
+#include "openturns/Bonmin.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/OSS.hxx"
+#include "openturns/SymbolicFunction.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#ifdef OPENTURNS_HAVE_BONMIN
+  #include "openturns/BonminProblem.hxx"
+  #include "BonBonminSetup.hpp"
+  #include "BonCbc.hpp"
+  using namespace Bonmin;
+  using namespace Ipopt;
+#endif
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(Bonmin)
+
+static const Factory<Bonmin> Factory_Bonmin;
+
+/** Constructors */
+Bonmin::Bonmin(const String & algoName)
+  : OptimizationAlgorithmImplementation()
+  , algoName_(algoName)
+  {
+    // Nothing to do
+  }   
+
+Bonmin::Bonmin( OptimizationProblem & problem,
+                const String & algoName)
+  : OptimizationAlgorithmImplementation(problem)
+  , algoName_()
+  {
+    setAlgorithmName(algoName);
+  }
+
+/* Virtual constructor */
+Bonmin * Bonmin::clone() const
+{
+  return new Bonmin(*this);
+}
+
+  
+  
+  /** Bonmin static methods */
+  Bool Bonmin::IsAvailable()
+  {
+    #ifdef OPENTURNS_HAVE_BONMIN
+      return true;
+    #else
+      return false;
+    #endif
+  }
+  
+  Description Bonmin::GetAlgorithmNames()
+  {
+    Description algoNames(5);
+    algoNames[0] = "B-BB";
+    algoNames[1] = "B-OA";
+    algoNames[2] = "B-QG";
+    algoNames[3] = "B-Hyb";
+    algoNames[4] = "B-iFP";
+    
+    return algoNames;
+  }
+
+  /** Accessors to Bonmin attributes */
+    // Algorithm name
+  void Bonmin::setAlgorithmName(const String & algoName)
+  {
+    // Check algoName
+    if (!GetAlgorithmNames().contains(algoName))
+    {
+      if (algoName == "B-Ecp")
+      {
+        throw InternalException(HERE) << "B-Ecp solver is unavailable.";
+      }
+      else
+      {
+        throw InvalidArgumentException(HERE) << "Unknown solver " << algoName;
+      }
+    }
+      
+    algoName_ = algoName;
+  }
+  
+  String Bonmin::getAlgorithmName() const
+  {
+    return algoName_;
+  }
+
+  
+/** Check whether this problem can be solved by this solver. */
+void Bonmin::checkProblem(const OptimizationProblem & problem) const
+{
+  // Cannot solve multi-objective problems
+  if (problem.hasMultipleObjective()) throw InvalidArgumentException(HERE) << "Error: Bonmin algorithm " << algoName_ << " does not support multi-objective optimization";
+  
+  // No LeastSquaresProblem / NearestPointProblem
+  if (problem.hasResidualFunction() || problem.hasLevelFunction())
+    throw InvalidArgumentException(HERE) << "Bonmin does not support least squares or nearest point problems";
+}
+
+
+#ifdef OPENTURNS_HAVE_BONMIN
+
+/** Accessors to Bonmin options */
+void getOptionsFromResourceMap(BonminSetup & bonmin)
+{
+//   Get options for Bonmin setup from ResourceMap
+//   See Bonmin/Ipopt user manuals for more details.
+  
+  std::vector<String> keys(ResourceMap::GetKeys());
+  UnsignedInteger nbKeys = keys.size();
+    
+  for (UnsignedInteger i=0; i<nbKeys; ++i)
```

Comment:
We should think about another solution here. And for the moment I do not see any specific key ?

Reply:
What is your concern here? Performance and efficiency? Or is it the fact we use the ResourceMap for this type of options? Or something else?

Regarding your second remark, indeed, only the user-defined options for Bonmin and Ipopt are stored into the ResourceMap. So no specific key is included by default.

Reply:
The current implementation suggests that the loop concerns **all** keys and we filter if string starts with "Bonmin". Currently we get around 500 keys.

It is not the purpose of this PR to solve the bad implementation of `ResourceMap`. However we might think about having an internal `Description` attribute that contains the list of keys for example or something that ease access to specific keys in order to avoid such loop.